_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
+
+For solving optimization problems, the Fréchet distance is not ideal, as it effectively measures the distance of a single point, the maximum error. One can informally describe an [L2] error metric as similar, except that instead of describing the maximum length of the leash, it describes an *effort* of handling the dog which is proportional to the square of the distance of the dog from the engineer. Minimizing an L2 error means minimizing that total effort. This error metric takes into account the entire path, and thus is smoother in response to small changes of the parameters.
+
+I believe there is no one perfect error metric, and the choice depends on the context. Fortunately, for smooth curves such as we're likely to find in fonts, the differences between these error metrics are subtle, and we can confidently choose the one that's easiest to reason about mathematically. For this blog post, that's the L2 norm.
+
+Two things more to say: for smooth curves + low error, leash is generally perpendicular. Also, normalized arc length (both engineer and Pierre move at a constant speed) is a good approximation, though strictly larger.
+
+## Stating the problem
+
+Our dog Pierre has a funny quirk: he is capable of moving only along a cubic bezier path. That said, he is otherwise fairly obedient, and always starts and ends at the same position as the engineer, as well as starting out and ending up in the same direction.
+
+Another way to state this is that our cubic bezier must match the source curve to G1. Then we want to find values for the parameters that minimizes the error norm. For a cubic bezier, the two parameters are the lengths of the two control arms; all other parameters are determined by the requirement to match the source curve.
+
+As a mathematical simplification, we can factor out translation, uniform scaling, and rotation, and just consider a curve that goes from (0, 0) to (1, 0), with given $\theta_0$ and $\theta_1$ angles.
+
+[image]
+
+Then, the control points of the bezier are $(0, 0)$, $(\delta_0 \cos \theta_0, \delta_0 \sin \theta_0)$, $(1 - \delta_1 \cos \theta_1, \delta_1 \sin \theta_1)$, $(1, 0)$.
+
+## Prior work
+
+Having stated the problem, we can take a look at the existing literature.
+
+* G2 geometric Hermite interpolation, de Boor. A bunch of new things here: a proof of O(n^6) scaling. The presence of up to three solutions. However, the constant factor on error is 5-10x, which is not great.
+
+* Two approaches from the Penner paper, center of mass and orthogonal distance fitting (ODF).
+
+* The Graphics Gems chapter.
+
+Also note, I addressed this problem in chapter 9 of my [thesis]. That approach gave good results but was very slow. The solution in this blog is strictly better.
```

Comment:
how do these compare to Potrace, btw? Since its whole thing was to be a good tradeoff solution (fast "enough" + accurate "enough") for cubic curve fitting.

Reply:
I added a short paragraph.